pkg/stream/stream.go

```diff
@@ -0,0 +1,160 @@
+func OpenKopiaRepository(ctx context.Context, password string) (repo.Repository, error) {
+	if _, err := os.Stat(defaultConfigFileName()); os.IsNotExist(err) {
+		return nil, errors.New("Failed find kopia configuration file")
+	}
+
+	r, err := repo.Open(ctx, defaultConfigFileName(), password, &repo.Options{})
+	if os.IsNotExist(err) {
+		return nil, errors.New("Failed to find kopia repository, use `kopia repository connect`")
+	}
+
+	return r, errors.Wrap(err, "Failed to open kopia repository")
+}
+
+func defaultConfigFileName() string {
+	return filepath.Join(os.Getenv("HOME"), ".config", "kopia", "repository.config")
+}
+
+// SnapshotSource creates and uploads a kopia snapshot to the given repository
+func SnapshotSource(ctx context.Context, rep repo.Repository, u *snapshotfs.Uploader, sourceInfo snapshot.SourceInfo, rootDir fs.Entry) error {
+	fmt.Printf("Snapshotting %v ...\n", sourceInfo)
+
+	t0 := time.Now()
+
+	previous, err := findPreviousSnapshotManifest(ctx, rep, sourceInfo, nil)
+	if err != nil {
+		return errors.Wrap(err, "Failed to find previous kopia manifests")
+	}
+
+	policyTree, err := policy.TreeForSource(ctx, rep, sourceInfo)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get kopia policy tree")
+	}
+
+	manifest, err := u.Upload(ctx, rootDir, policyTree, sourceInfo, previous...)
+	if err != nil {
+		return errors.Wrap(err, "Failed to upload the kopia snapshot")
+	}
+
+	manifest.Description = snapshotDescription
+
+	snapID, err := snapshot.SaveSnapshot(ctx, rep, manifest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to save kopia manifest")
+	}
+
+	_, err = policy.ApplyRetentionPolicy(ctx, rep, sourceInfo, true)
+	if err != nil {
+		return errors.Wrap(err, "Failed to apply kopia retention policy")
+	}
+
+	if ferr := rep.Flush(ctx); ferr != nil {
+		return errors.Wrap(ferr, "Failed to flush kopia repository")
+	}
+
+	fmt.Printf("\nCreated snapshot with root %v and ID %v in %v\n", manifest.RootObjectID(), snapID, time.Since(t0).Truncate(time.Second))
+
+	return err
+}
+
+// findPreviousSnapshotManifest returns the list of previous snapshots for a given source,
+// including last complete snapshot
+func findPreviousSnapshotManifest(ctx context.Context, rep repo.Repository, sourceInfo snapshot.SourceInfo, noLaterThan *time.Time) ([]*snapshot.Manifest, error) {
+	man, err := snapshot.ListSnapshots(ctx, rep, sourceInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to list previous kopia snapshots")
+	}
+
+	// find latest complete snapshot
+	var previousComplete *snapshot.Manifest
+	var result []*snapshot.Manifest
+
+	for _, p := range man {
+		if noLaterThan != nil && p.StartTime.After(*noLaterThan) {
+			continue
+		}
+
+		if p.IncompleteReason == "" && (previousComplete == nil || p.StartTime.After(previousComplete.StartTime)) {
+			previousComplete = p
+		}
+	}
+
+	if previousComplete != nil {
+		result = append(result, previousComplete)
```

Comment:
I found this a little confusing, but I see what you did. This is a slice with zero or one entries, which you later use as a variadic parameter